src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,298 @@
+<?php
+
+namespace M6Web\Tornado\Adapter\Swoole;
+
+use Generator;
+use M6Web\Tornado\Adapter\Swoole\Internal\DummyPromise;
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\Promise;
+use Swoole\Coroutine;
+use RuntimeException;
+use Swoole\IDEHelper\StubGenerators\Swoole;
+use Swoole\Runtime;
+use Throwable;
+use function extension_loaded;
+
+class EventLoop implements \M6Web\Tornado\EventLoop
+{
+    private $cids;
+
+    public function __construct()
+    {
+        if (!extension_loaded('swoole')) {
+            throw new RuntimeException(
+                'EventLoop must running only with swoole extension.'
+            );
+        }
+
+        $this->cids = [];
+    }
+
+    private function shiftCoroutine(): mixed
+    {
+        if(count($this->cids) === 0) {
+            return null;
+        }
+
+        $cid = array_shift($this->cids);
+        if(Coroutine::exists($cid)) {
+            Coroutine::resume($cid);
+        } else {
+            $this->shiftCoroutine();
+        }
+
+        return $cid;
+    }
+
+    private function pushCoroutine(): mixed
+    {
+        $cid = Coroutine::getCid();
+        $this->cids[] = $cid;
+        Coroutine::yield();
+        return $cid;
+    }
+
+    private function createPromise(): DummyPromise
+    {
+        return new DummyPromise(function () {
+            $this->shiftCoroutine();
+        });
+    }
+
+    private function resolve($value)
+    {
+        if($value instanceof DummyPromise && !$value->isPending()) {
+            if($value->getException() !== null) {
+                throw $value->getException();
+            }
+
+            $value = $this->resolve($value->getValue());
+        }
+
+        if(is_array($value)) {
+            foreach ($value as $k => $v) {
+                $value[$k] = $this->resolve($v);
+            }
+        }
+
+        return $value;
```

Comment:
Sorry, I don't understand this while function… it's recursive, but also related to `array`… I don't see the point 😕

Reply:
`resolve` is involving in `wait` and `async` function.
I agree the recursive function shouldn't be specific to types.
I didn't find a smart way to get the resolution at the origin and propagate it's values.

Note that `array` is involved in `promiseAll` and return an `array` of promise. This may justify the `array` type check.